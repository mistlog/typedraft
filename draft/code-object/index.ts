export * from "./export-class";
export * from "./inline-context";
export * from "./local-context";
export * from "./method";
export * from "./module";
