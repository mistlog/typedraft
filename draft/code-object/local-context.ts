import { FunctionDeclaration, Statement } from "@babel/types";
import { Binding, NodePath } from "@babel/traverse";
import { IDSL } from "../core/transcriber";

export interface ILocalContext {
    ToStatements: () => Array<Statement>;
}

export class LocalContext {
    m_Binding: Binding;

    /**
     * binding is used to find all references to this context, thus we can replace them with resolved statements
     */
    constructor(binding: Binding) {
        this.m_Binding = binding;
    }

    get m_Path() {
        return this.m_Binding.path as NodePath<FunctionDeclaration>;
    }

    get m_Code() {
        return this.m_Path.node;
    }

    get m_Refs() {
        return this.m_Binding.referencePaths;
    }

    get m_Name() {
        return this.m_Code.id.name;
    }

    /**
     * The context itself does nothing but expresses our intention. If it's used with a DSL, we can resolve it to get the "real" statements.
     */
    Resolve(dsl: IDSL) {
        /**
         * remove all directives, because we don't need "use ..." after resolved
         */
        this.m_Code.body.directives = [];
        this.m_Code.body.body = dsl.Transcribe(this.ToStatements(), this.m_Path);
    }

    ToStatements() {
        return this.m_Code.body.body;
    }

    GetDSLName() {
        const [directive] = this.m_Code.body.directives;

        if (directive) {
            const [use, dsl_name] = directive.value.value.split(" ");
            return dsl_name;
        }

        return "";
    }
}
