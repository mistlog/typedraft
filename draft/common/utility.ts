import { template } from "@babel/core";
import { Node } from "@babel/types";
import { NodePath } from "@babel/traverse";
import { default as Binding } from "@babel/traverse/lib/scope/binding";
import { parse } from "@babel/parser";
import generate, { GeneratorOptions } from "@babel/generator";

export function ToAst<T = any>(code: string) {
    return (template.ast(code, { plugins: ["typescript", "jsx"] }) as unknown) as T;
}

export function ToNodePath<T = any>(code: string): NodePath<T> {
    const path = new NodePath<T>(null, null);
    path.node = (ToAst(code) as unknown) as T;
    return path;
}

/**
 * only used in test to construct required binding param
 */
export function ToBinding(code: string): Binding {
    const binding = new Binding({
        path: ToNodePath<Node>(code),
        existing: null,
        identifier: null,
        scope: null,
        kind: "let",
    });
    return binding;
}

export function ToFile(raw: string) {
    // babel bug: https://github.com/babel/babel/issues/8837
    // refert to test case "transcriber.interface-no-parse-error"
    const code = raw.replace(new RegExp("^<", "gm"), ";<");
    return parse(code, {
        sourceType: "module",
        plugins: ["typescript", "jsx", ["decorators", { decoratorsBeforeExport: true }]],
    });
}

export function ToString(node: Node, options?: GeneratorOptions) {
    return generate(node, options || {}).code;
}
