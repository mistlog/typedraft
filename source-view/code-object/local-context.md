# Local Context

The concept of local context is essential in typedraft, any function declaration is treated as a local context.

```typescript
export class LocalContext {
    m_Code: FunctionDeclaration;
    m_Binding: Binding | null;
    m_Path: NodePath<FunctionDeclaration>;
    get m_Refs() {
        return this.m_Binding.referencePaths;
    }
    get m_Name() {
        return this.m_Code.id.name;
    }
}
```

The context itself does nothing but expresses our intention. If it's used with a DSL, we can resolve it to get the "real" statements.

```typescript
<LocalContext /> +
    function Resolve(this: LocalContext & ILocalContext, dsl: IDSL) {
        // because we don't need "use ..." after resolved
        this.m_Code.body.directives = [];
        this.m_Code.body.body = dsl.Transcribe(this.ToStatements(), this.m_Path);
    };
```

## Example in DSL
For example, if we want to use SQL as a DSL in typescript, we can do it in this way. Suppose that we want to select some data from an array:

```typescript
const array = [
    {name:"cpp",version:17},
    {name:"c",version:99},
    {name:"js",version:6}
]

// expected result: ["c"]
const data = array
    .filter(each=>each.name.startsWith("c") && each.version > 20)
    .map(each=>each.name);
```

In this context, we can write it as:

```typescript
function SelectData(){
    "use sql"
    
    const data = `
        select ${name} 
        from ${array} 
        where ${name.startsWith("c") && version > 20}
    `;
}
```
You may find that name, version and array are not defined. To make the compiler happy, you can add

```typescript
let name: string;
let version: number;
...
```

just above ```const data=...```. We can always find a way to both enjoy the convenience of context and leverage the power of typescript. That depends on the genius of the DSL designer.

The crux here is that as long as we have enough infomation to express our intention in this point(context), we can translate it into equvalent typescript.

## Example in Literate Programming

Suppose that we want to use that data:

```typescript
const array = [
    {name:"cpp",version:17},
    {name:"c",version:99},
    {name:"js",version:6}
]

// expected result: ["c"]
const data = array
    .filter(each=>each.name.startsWith("c") && each.version > 20)
    .map(each=>each.name);

console.log(data);
```

With context, this piece of code can be written as:

```typescript
const array = [
    {name:"cpp",version:17},
    {name:"c",version:99},
    {name:"js",version:6}
]

<SelectData/>;
<UseData/>;
```

We use JSX syntax here and the tag name is the function name. The UseData context could be:

```typescript
function UseData(){
    console.log(data)
}
```

We don't use any DSL, then we just use the "host" language and the output(statements) would be just the input. Besides, as functions can be placed in any place in a source file, we can rearrange code now.

# Utility

```typescript
<LocalContext /> +
    function ToStatements(this: LocalContext) {
        return this.m_Code.body.body;
    };
```

```typescript
<LocalContext /> +
    function GetContextName(this: LocalContext) {
        const [directive] = this.m_Code.body.directives;
        if (directive) {
            const [, context_name] = directive.value.value.split(" ");
            return context_name;
        }
        return "";
    };
```

# Trivial

```typescript
<LocalContext /> +
    function constructor(this: LocalContext, node: FunctionDeclaration, binding?: Binding, path?: NodePath<FunctionDeclaration>) {
        this.m_Code = node;

        // binding is used to find all references to this context, thus we can replace them with resolved statements
        this.m_Binding = binding || null;
        this.m_Path = path;
    };
```

```typescript
export interface ILocalContext {
    ToStatements: () => Array<Statement>;
}
```