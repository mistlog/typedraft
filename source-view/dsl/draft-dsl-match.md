```typescript
export class PatternMatch {}
```

```typescript
<PatternMatch /> +
    function Transcribe(block: Array<ExpressionStatement>): [IfStatement] {
        let tail: IfStatement = null;
        let head: IfStatement = null;
        block.forEach(each => {
            const expression = each.expression as ArrowFunctionExpression;
            <TranscribeExpressionToIf />;
        });

        /**
         * transcribed is the "head"(first) of if statement
         */
        return [head];
    };
```

```typescript
function TranscribeExpressionToIf(expression: ArrowFunctionExpression, head: IfStatement, tail: IfStatement) {
    if (expression.params.length === 0) {
        <HandleDefaultCase />;
    } else {
        <BuildCurrentIf />;
        <MoveToNext />;
    }
}
```

```typescript
function BuildCurrentIf(expression: ArrowFunctionExpression) {
    //
    const [pattern_info] = expression.params as [Identifier];
    const to_match = pattern_info.name;
    const annotation = (pattern_info.typeAnnotation as TypeAnnotation).typeAnnotation;

    //
    let current: IfStatement = null;
    if (isTSLiteralType(annotation)) {
        /**
         * number or string
         */
        type Literal = (StringLiteral | NumberLiteral) & { extra: { raw: string } };
        const pattern = (annotation.literal as Literal).extra.raw;
        current = ToAst(`if(${to_match}===${pattern}){}`) as IfStatement;
    } else if (isTSTypeReference(annotation)) {
        /**
         * enum
         */
        current = ToAst(`if(${to_match}===${ToString(annotation)}){}`) as IfStatement;
    }
    current.consequent = expression.body as BlockStatement;
}
```

```typescript
function MoveToNext(head: IfStatement, current: IfStatement, tail: IfStatement) {
    if (head == null) {
        head = tail = current;
    } else {
        tail = tail.alternate = current;
    }
}
```

```typescript
function HandleDefaultCase(expression: ArrowFunctionExpression, tail: IfStatement) {
    tail.alternate = expression.body as BlockStatement;
}
```