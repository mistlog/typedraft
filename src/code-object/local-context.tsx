/**
 * # Local Context
 */
export class LocalContext {
    m_Binding: Binding;

    get m_Path() {
        return this.m_Binding.path as NodePath<FunctionDeclaration>;
    }
    get m_Code() {
        return this.m_Path.node;
    }
    get m_Refs() {
        return this.m_Binding.referencePaths;
    }
    get m_Name() {
        return this.m_Code.id.name;
    }
}

/**
 * The context itself does nothing but expresses our intention. If it's used with a DSL, we can resolve it to get the "real" statements.
 */
<LocalContext /> +
    function Resolve(this: LocalContext & ILocalContext, dsl: IDSL) {
        /**
         * remove all directives, because we don't need "use ..." after resolved
         */
        this.m_Code.body.directives = [];
        this.m_Code.body.body = dsl.Transcribe(this.ToStatements(), this.m_Path);
    };

<LocalContext /> +
    function ToStatements(this: LocalContext) {
        return this.m_Code.body.body;
    };

<LocalContext /> +
    function GetDSLName(this: LocalContext) {
        const [directive] = this.m_Code.body.directives;
        if (directive) {
            const [use, dsl_name] = directive.value.value.split(" ");
            return dsl_name;
        }
        return "";
    };

/**
 * # Trivial
 */
<LocalContext /> +
    function constructor(this: LocalContext, binding: Binding) {
        /**
         * binding is used to find all references to this context, thus we can replace them with resolved statements
         */
        this.m_Binding = binding;
    };

export interface ILocalContext {
    ToStatements: () => Array<Statement>;
}

import { FunctionDeclaration, Statement } from "@babel/types";
import { Binding, NodePath } from "@babel/traverse";
import { IDSL } from "../core/transcriber";
