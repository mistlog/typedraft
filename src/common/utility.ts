import { template } from "@babel/core";
import { Node } from "@babel/types";
import generate, { GeneratorOptions } from "@babel/generator";
import { parse } from "@babel/parser";

export function ToAst(code: string)
{
    return template.ast(code, { plugins: ["typescript", "jsx"] });
}

export function ToFile(code: string)
{
    return parse(code, { sourceType: "module", plugins: ["typescript", "jsx"] });
}

export function ToString(node: Node, options?: GeneratorOptions)
{
    return generate(node, options || {}).code;
}