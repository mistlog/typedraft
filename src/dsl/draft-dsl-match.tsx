import { Identifier, TypeAnnotation, isTSLiteralType, IfStatement, BlockStatement, StringLiteral, NumberLiteral, isTSTypeReference, ArrowFunctionExpression, Statement, isArrowFunctionExpression, ExpressionStatement, ifStatement } from "@babel/types";
import { ToAst, ToString } from "../common/utility";

export class PatternMatch
{

}

<PatternMatch /> + function Transcribe(block: Array<ExpressionStatement>): [IfStatement]
{
    let tail: IfStatement = null;
    let head: IfStatement = null;

    block.forEach(each =>
    {
        const expression = each.expression as ArrowFunctionExpression;

        //@ts-ignore
        <TranscribeExpressionToIf />;
    })


    /**
     * transcribed is the "head"(first) of if statement
     */
    return [head];
}

function TranscribeExpressionToIf(expression: ArrowFunctionExpression, head: IfStatement, tail: IfStatement)
{
    if (expression.params.length === 0)
    {
        //@ts-ignore
        <HandleDefaultCase />
    }
    else
    {
        //@ts-ignore
        <BuildCurrentIf />;

        //@ts-ignore
        <MoveToNext />;
    }
}

function BuildCurrentIf(expression: ArrowFunctionExpression)
{
    //
    const [pattern_info] = expression.params as [Identifier];
    const to_match = pattern_info.name;
    const annotation = (pattern_info.typeAnnotation as TypeAnnotation).typeAnnotation;

    //
    let current: IfStatement = null;

    if (isTSLiteralType(annotation))
    {
        /**
         * number or string
         */
        type Literal = (StringLiteral | NumberLiteral) & { extra: { raw: string } };
        const pattern = (annotation.literal as Literal).extra.raw;
        current = ToAst(`if(${to_match}===${pattern}){}`) as IfStatement;
    }
    else if (isTSTypeReference(annotation))
    {
        /**
         * enum
         */
        current = ToAst(`if(${to_match}===${ToString(annotation)}){}`) as IfStatement;
    }

    current.consequent = expression.body as BlockStatement;
}

function MoveToNext(head: IfStatement, current: IfStatement, tail: IfStatement)
{
    if (head == null)
    {
        head = tail = current;
    }
    else
    {
        tail = tail.alternate = current;
    }

}


function HandleDefaultCase(expression: ArrowFunctionExpression, tail: IfStatement)
{
    tail.alternate = expression.body as BlockStatement;
}