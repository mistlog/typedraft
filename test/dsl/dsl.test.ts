import { ToAst, ToString } from "../../src/common/utility";
import { FunctionDeclaration } from "@babel/types";
import { LocalContext } from "../../src/code-object/local-context";
import { PatternMatch } from "../../src/dsl/draft-dsl-match";

describe("dsl", () =>
{
    test("dsl.match.number", () =>
    {
        //
        const code = `
            function Test(value:number){
                'use match';

                (value: 1) =>
                {
                    console.log(1);
                }
            
                (value: 2) =>
                {
                    console.log(2);
                }
            
                (value: 3) =>
                {
                    console.log(3);
                }
            }
        `;

        const context = new LocalContext(ToAst(code) as FunctionDeclaration);
        context.Resolve(new PatternMatch());
        expect(ToString(context.m_Code)).toMatchSnapshot();
    })

    test("dsl.match.string", () =>
    {
        //
        const code = `
            function Test(value:string){
                'use match';

                (value: "a") =>
                {
                    console.log("a");
                }
            
                (value: "b") =>
                {
                    console.log("b");
                }
            }
        `;

        const context = new LocalContext(ToAst(code) as FunctionDeclaration);
        context.Resolve(new PatternMatch());
        expect(ToString(context.m_Code)).toMatchSnapshot();

    })

    test("dsl.match.enum", () =>
    {
        //
        const code = `
            function Test(value:Event){
                'use match';

                (value: Event.EventA) =>
                {
                    console.log(1);
                }
            
                (value: Event.EventB) =>
                {
                    console.log("a");
                }
            }
        `;

        const context = new LocalContext(ToAst(code) as FunctionDeclaration);
        context.Resolve(new PatternMatch());
        expect(ToString(context.m_Code)).toMatchSnapshot();

    })

    test("dsl.match.only-default: only default is not allowed", () =>
    {
        //
        const code = `
            function Test(value:number){
                'use match';

                () =>
                {
                    console.log(2);
                }
            }
        `;

        const context = new LocalContext(ToAst(code) as FunctionDeclaration);
        expect(() => context.Resolve(new PatternMatch())).toThrowError();
    })

    test("dsl.match.default", () =>
    {
        //
        const code = `
            function Test(value:number){
                'use match';

                (value: 1) =>
                {
                    console.log(1);
                }
            
                () =>
                {
                    console.log(2);
                }
            }
        `;

        const context = new LocalContext(ToAst(code) as FunctionDeclaration);
        context.Resolve(new PatternMatch());
        expect(ToString(context.m_Code)).toMatchSnapshot();
    })
})