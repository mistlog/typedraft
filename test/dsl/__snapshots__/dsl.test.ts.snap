// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`dsl dsl.match.default 1`] = `
"function Test(value: number) {
  if (value === 1) {
    console.log(1);
  } else {
    console.log(2);
  }
}"
`;

exports[`dsl dsl.match.enum 1`] = `
"function Test(value: Event) {
  if (value === Event.EventA) {
    console.log(1);
  } else if (value === Event.EventB) {
    console.log(\\"a\\");
  }
}"
`;

exports[`dsl dsl.match.number 1`] = `
"function Test(value: number) {
  if (value === 1) {
    console.log(1);
  } else if (value === 2) {
    console.log(2);
  } else if (value === 3) {
    console.log(3);
  }
}"
`;

exports[`dsl dsl.match.string 1`] = `
"function Test(value: string) {
  if (value === \\"a\\") {
    console.log(\\"a\\");
  } else if (value === \\"b\\") {
    console.log(\\"b\\");
  }
}"
`;
